### PR TITLE
ReleaseShellContext() and ReleaseAllShellContexts() methods

### PR DESCRIPTION
**Pull request by jtkech**

### CODE REVIEW

**Comment by sebastienros** [comment]
why not add a new TenantState, Reloading ...
Also can you explain what scenario it's fixing?

**Comment by jtkech** [comment]
- Originally it was just to add the ability, as we discussed, to release one or multiples tenants to free up resources e.g after some inactivities. Then working on it i saw the following.

- Before, a released shell was always removed from the list, now this is not always the case when a dependent shell is released, when pre-creating all shells by only adding place holders (marked as released). This allows to lazily built a shell while being able to get its settings even it is not yet built.

    **So here, one goal is that the shells list always contains all shells so that e.g `GetAllSettings()` always works, at least the most little time without a given shell in the list**.

- Following this idea, when a feature is enabled / disabled we still 1st remove the shell from the list but here, **in place of waiting the deferred task, we re-add a place holder in the list immediately**.

- Idem when reloading a tenant, here we re-add immediately a `ReloadedPlaceHolder`. So here `Reloading` **is more a placeholder state than a tenant state that would need to be persisted**, but maybe better to make it a placeholder state (as did at some point) in place of using a derived class.

- `CreateShellContextAsync` private, was only used on setup where we can use `GetScopeAsync`.

- When updating settings **that are not shell settings**, we don't need  `ReloadShellContextAsync()` that reloads settings, now we can just use `ReleaseShellContextAsync()`.

**Comment by jtkech** [comment]
@sebastienros just for infos as a new summary for this PR.

- Adds `ReleaseShellContextAsync()` to free up resources while still serving the tenant, and through extensions `ReleaseAllShellContextAsync()` and `ReloadAllShellContextsAsync()`.

- After removing a shell from the list, re-add immediately a place holder so that its settings can be retrieved even the shell is not yet rebuilt.

- **I removed the `Loading` state**, after upating the settings, i found another way to reload them in a subsequent scope, this by calling `ReloadShellContextAsync()` itself in a deferred task.

- `CreateShellContextAsync()` private, was only used on setup where we can use `GetScopeAsync()`.

- Use `ReleaseShellContextAsync()` in place of `ReloadShellContextAsync()` everywhere we are updating settings that in fact are not shell settings.

- Fixes other little issues.

File: src/OrchardCore/OrchardCore.Abstractions/Shell/ShellExceptions.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+using System;
+
+namespace OrchardCore.Environment.Shell
+{
+    public static class ShellExceptions
+    {
+        public class ConcurrencyException : Exception
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Abstractions/Shell/ShellExceptions.cs]: No inner class for an exception

File: src/OrchardCore/OrchardCore/Environment/Shell/ShellHost.cs [hunk 2]
```diff
@@ -164,8 +164,11 @@ public async Task ReloadShellContextAsync(ShellSettings settings)
                 settings = await _shellSettingsManager.LoadSettingsAsync(settings.Name);
             }
 
-            while (true)
+            var count = 0;
+            while (count < 9)
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore/Environment/Shell/ShellHost.cs]: Make it a constant, like `MaxComethingAttempts`

**Comment by jtkech** [comment]
Yes, maybe better to wait after rc2 for distributed cache and redis services.

But i will merge this one as it already fixes some race conditions.

**Pull request merged**